### PR TITLE
Add a negative test for ingress.class handling

### PR DESCRIPTION
**Pull request by nak3**
This patch adds a negative test for ingress.class handling.

The new TestIngressClass test verifies that kingrress does not picck
ingress up when ingress.class annotation is incorrect.

Fixes https://github.com/knative/networking/issues/142

/cc @mattmoor @tcnghia @ZhiminXiang

### CODE REVIEW

File: test/conformance/ingress/class.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	},
+		// Override ingress.class annotation
+		func(ing *v1alpha1.Ingress) {
+			ing.Annotations = class
+		},
+	)
+
+	const (
+		interval = 2 * time.Second
+		duration = 6 * time.Second
```
Comment (mattmoor) [inline test/conformance/ingress/class.go:91]: I'd give it 30s

File: test/conformance/ingress/class.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	for {
+		select {
+		case <-ticker.C:
+			var ing *v1alpha1.Ingress
+			err := reconciler.RetryTestErrors(func(attempts int) (err error) {
+				ing, err = clients.NetworkingClient.Ingresses.Get(ctx, org.Name, metav1.GetOptions{})
+				return err
+			})
+			if err != nil {
+				cancel()
```
Comment (mattmoor) [inline test/conformance/ingress/class.go:103]: The `cancel` is already set up with `t.Cleanup`, you only need to call cancel if you want things torn down BEFORE the end of the test.

File: test/conformance/ingress/class.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+func TestIngressClass(t *testing.T) {
+	t.Parallel()
+	ctx, clients := context.Background(), test.Setup(t)
+
+	// Create a backend service to create valid ingress except for invalid ingress.class.
+	name, port, _ := CreateRuntimeService(ctx, t, clients, networking.ServicePortNameHTTP1)
+
+	tests := []struct {
+		name  string
+		class map[string]string
```
Comment (tcnghia) [inline test/conformance/ingress/class.go:43]: Rename to `annotations`?

File: test/conformance/ingress/class.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			createIngressWithClass(ctx, t, clients, test.class, name, port)
+		})
+	}
+
+}
+
+func createIngressWithClass(ctx context.Context, t *testing.T, clients *test.Clients, class map[string]string, name string, port int) {
```
Comment (tcnghia) [inline test/conformance/ingress/class.go:63]: You could pass `IngressBackend` instead of `(name,port)`

**Review by tcnghia — COMMENTED** [review]
Thanks for the tests!

File: test/conformance/ingress/class.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+	// Create a backend service to create valid ingress except for invalid ingress.class.
+	name, port, _ := CreateRuntimeService(ctx, t, clients, networking.ServicePortNameHTTP1)
+
+	tests := []struct {
+		name  string
+		class map[string]string
+	}{{
+		name:  "ommited",
+		class: map[string]string{},
```
Comment (vagababov) [inline test/conformance/ingress/class.go:46]: you should also add a case with nil map.

File: test/conformance/ingress/class.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+						},
+					}},
+				}},
+			},
+		}},
+	},
+		// Override ingress.class annotation
+		func(ing *v1alpha1.Ingress) {
+			ing.Annotations = class
+		},
```
Comment (vagababov) [inline test/conformance/ingress/class.go:86]: layout seems funky?

File: test/conformance/ingress/class.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+				ing, err = clients.NetworkingClient.Ingresses.Get(ctx, org.Name, metav1.GetOptions{})
+				return err
+			})
+			if err != nil {
+				cancel()
+				t.Fatal("Error getting Ingress:", err)
+			}
+			// Verify ingress is not changed.
+			if !equality.Semantic.DeepEqual(org, ing) {
+				t.Fatalf("Unexpected update, want=%v, got=%v", org, ing)
```
Comment (vagababov) [inline test/conformance/ingress/class.go:108]: Got before want.
```suggestion
				t.Fatalf("Update = %v, want = %v, diff(-want;+got)\n%s", ing, org, cmp.Diff(org, ing))
```

File: test/conformance/ingress/class.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+			err := reconciler.RetryTestErrors(func(attempts int) (err error) {
+				ing, err = clients.NetworkingClient.Ingresses.Get(ctx, org.Name, metav1.GetOptions{})
+				return err
+			})
+			if err != nil {
+				cancel()
+				t.Fatal("Error getting Ingress:", err)
+			}
+			// Verify ingress is not changed.
+			if !equality.Semantic.DeepEqual(org, ing) {
```
Comment (vagababov) [inline test/conformance/ingress/class.go:107]: `cmp.Equal`?

File: test/conformance/ingress/class.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+			createIngressWithClass(ctx, t, clients, test.class, name, port)
+		})
+	}
+
+}
+
+func createIngressWithClass(ctx context.Context, t *testing.T, clients *test.Clients, class map[string]string, name string, port int) {
+	t.Helper()
+
+	org, cancel := CreateIngress(ctx, t, clients, v1alpha1.IngressSpec{
```
Comment (vagababov) [inline test/conformance/ingress/class.go:66]: s/org/original/

**Comment by nak3** [comment]
Thank you for the review! Updated.

File: test/conformance/ingress/class.go [hunk 10]
```diff
@@ -37,75 +37,76 @@ func TestIngressClass(t *testing.T) {
...
 	tests := []struct {
-		name  string
-		class map[string]string
+		name        string
+		annotations map[string]string
 	}{{
-		name:  "ommited",
-		class: map[string]string{},
+		name:        "nil",
+		annotations: nil,
```
Comment (vagababov) [inline test/conformance/ingress/class.go:51]: this can simply be eluded, but up to you.
```suggestion
```

**Review by vagababov — COMMENTED** [review]
/lgtm
/hold
for others

File: test/conformance/ingress/class.go [hunk 11]
```diff
@@ -37,75 +37,76 @@ func TestIngressClass(t *testing.T) {
...
 			})
 			if err != nil {
-				cancel()
 				t.Fatal("Error getting Ingress:", err)
 			}
 			// Verify ingress is not changed.
-			if !equality.Semantic.DeepEqual(org, ing) {
-				t.Fatalf("Unexpected update, want=%v, got=%v", org, ing)
+			if !cmp.Equal(original, ing) {
+				t.Fatalf("Update = %v, want = %v, diff(-want;+got)\n%s", ing, original, cmp.Diff(original, ing))
```
Comment (mattmoor) [inline test/conformance/ingress/class.go:108]: generally just the diff is enough

**Review by mattmoor — COMMENTED** [review]
/lgtm

File: test/conformance/ingress/class.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/google/go-cmp/cmp"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"knative.dev/networking/pkg/apis/networking"
+	"knative.dev/networking/pkg/apis/networking/v1alpha1"
+	"knative.dev/networking/test"
+	"knative.dev/pkg/reconciler"
+)
+
+// TestIngressClass verifies that kingrress does not picck ingress up when ingress.class annotation is incorrect.
```
Comment (ZhiminXiang) [inline test/conformance/ingress/class.go:33]: nit: picck -> pick

File: test/conformance/ingress/class.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+		name:        "incorrect",
+		annotations: map[string]string{networking.IngressClassAnnotationKey: "incorrect"},
+	}, {
+		name:        "empty",
+		annotations: map[string]string{networking.IngressClassAnnotationKey: ""},
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			createIngressWithAnnotations(ctx, t, clients, test.annotations, ingressBackend)
```
Comment (ZhiminXiang) [inline test/conformance/ingress/class.go:64]: I would rename the function name as `verifyIngressWithAnnotations`. WDYT?

Comment (nak3) [reply test/conformance/ingress/class.go]: Yes, sounds good. Thank you. Updated.

**Pull request merged**